src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -411,63 +491,91 @@ void LeaderWatcher<I>::handle_get_locker(int r,
 
   Mutex::Locker timer_locker(m_threads->timer_lock);
   Mutex::Locker mutex_locker(m_lock);
+  assert(!m_timer_op_tracker.empty());
 
   if (m_leader_lock->is_shutdown()) {
     dout(20) << "canceling due to shutdown" << dendl;
+    m_timer_op_tracker.finish_op();
     return;
   }
 
   if (is_leader(m_lock)) {
     m_locker = {};
-  } else {
-    if (r == -ENOENT) {
-      acquire_leader_lock(true);
-    } else {
-      if (r < 0) {
-        derr << "error retrieving leader locker: " << cpp_strerror(r) << dendl;
-      } else {
-        m_locker = locker;
-      }
+    m_timer_op_tracker.finish_op();
+    return;
+  }
 
-      schedule_timer_task("acquire leader lock",
-                          m_cct->_conf->rbd_mirror_leader_max_missed_heartbeats,
-                          false, &LeaderWatcher<I>::acquire_leader_lock);
+  if (r == -ENOENT) {
+    m_locker = {};
+    m_acquire_attempts = 0;
+    acquire_leader_lock();
+    return;
+  } else if (r < 0) {
+    derr << "error retrieving leader locker: " << cpp_strerror(r) << dendl;
+    schedule_get_locker(true, 1);
+    m_timer_op_tracker.finish_op();
+    return;
+  }
+
+  if (m_locker != locker) {
+    m_locker = locker;
+    if (m_acquire_attempts > 1) {
+      dout(10) << "new lock owner detected -- resetting heartbeat counter"
+               << dendl;
+      m_acquire_attempts = 0;
     }
   }
-}
 
-template <typename I>
-void LeaderWatcher<I>::acquire_leader_lock() {
-    return acquire_leader_lock(false);
+  if (m_acquire_attempts >
+        m_cct->_conf->rbd_mirror_leader_max_acquire_attempts_before_break) {
```

Comment:
@dillaman Shouldn't it be `>=`? 
"max_acquire_attempts" suggests that we don't attempt more than this, while now we attempt one more time than this. TestLeaderWatcher.Break now runs for 40 sec instead of 30 (with the message "breaking leader lock after 4 failed attempts to acquire" in logs).